File: algorithm/257.二叉树的所有路径.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 * 思路2: 看 c嘎嘎 版
 */
var binaryTreePaths = function(root) {
  const res = [];
  if (!root) return res;
  function _binaryTreePaths(node, str) {
    if (!node) {

    } else if (!node.left && !node.right) {
      res.push(str);
    } else {
      if (node.left) {
        _binaryTreePaths(node.left, `${str}->${node.left.val}`);
      }
      if (node.right) {
        _binaryTreePaths(node.right, `${str}->${node.right.val}`);
      }
    }
  }
  _binaryTreePaths(root, `${root.val}`);
  return res;
};

File: algorithm/257.二叉树的所有路径.cpp
#include "iostream"
#include "string"
#include "vector"
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/**
 * 思路:
 *  每次执行函数都创建一个数组
 *  然后递归地得到它子节点的路径
 *  之后循环一下连接起来
 */
class Solution {
public:
  vector<string> binaryTreePaths(TreeNode* root) {
    vector<string> res;
    if(!root) return res;
    if(!root -> left && !root -> right) {
      res.push_back(to_string(root -> val));
      return res;
    }
    vector<string> leftRes = binaryTreePaths(root -> left);
    for(unsigned i = 0; i < leftRes.size(); i++) {
      res.push_back(to_string(root -> val) + "->" + leftRes[i]);
    }
    vector<string> rightRes = binaryTreePaths(root -> right);
    for(unsigned i = 0; i < rightRes.size(); i++) {
      res.push_back(to_string(root -> val) + "->" + rightRes[i]);
    }
    return res;
  }
};
